policy: Add selectorcache cardinality metrics

joestringer opened the pull request:
Add metrics to track the maximum cardinality of selectors actively used
to select peers in the cluster. This can be useful as a hint as to the
use of policy rules that are overly permissive, leading to high policy
map usage.

Example with some sample policies + usage:

```
  # cilium-dbg policy selectors
  SELECTOR                                                                                                                                                                       LABELS                              USERS   IDENTITIES
  &LabelSelector{...}   default/from-world-to-role-public   2       63484
  &LabelSelector{...}   default/fqdn                        1       39546
  &LabelSelector{...}   default/rule1                       1       4992
                                                                    39546
                                                                    63484
  &LabelSelector{...}   default/fqdn                        3       45238
  &LabelSelector{...}   default/fqdn                        1
  &LabelSelector{...}   default/from-world-to-role-public   1       9
                                                                    16777217
  &LabelSelector{...}   default/from-world-to-role-public   1       10
  &LabelSelector{...}   default/from-world-to-role-public   1       2
  api.github.com ...    default/fqdn                        1       16777217
  # cilium-dbg metrics list | grep selector
  cilium_fqdn_selectors                                     1.000000
  cilium_policy_selector_match_count_max   class="cluster"  0.000000
  cilium_policy_selector_match_count_max   class="fqdn"     1.000000
  cilium_policy_selector_match_count_max   class="other"    3.000000
  cilium_policy_selector_match_count_max   class="world"    2.000000
```

The cilium_fqdn_selectors metric counts the total number of fqdn
selectors altogether, but the cilium_policy_selector_match_count_max is
the maximum number of identities selected by any selector in the
specified class.

For now to avoid high cardinality for the metric itself, there are four "classes":
- World: Equivalent to entity selector on world
- Cluster: Equivalent to entity selector on the cluster
- FQDN: Any selector that is based on an FQDNs statement
- Other: All other selectors.

"All" is notably omitted, because the Cilium dataplane already optimizes
such statements to minimize cardinality when encoding the identities into
policy maps.

Review commit by commit.

```
Add new metric "cilium_policy_selector_match_count_max" which counts the maximum number of identities selected by a selector of various different selector classes.
```

joestringer committed:
policy: Add selectorcache cardinality metrics

Add metrics to track the maximum cardinality of selectors actively used
to select peers in the cluster. This can be useful as a hint as to the
use of policy rules that are overly permissive, leading to high policy
map usage.

Example with some sample policies + usage:

  # cilium-dbg policy selectors
  SELECTOR                                                                                                                                                                       LABELS                              USERS   IDENTITIES
  &LabelSelector{...}   default/from-world-to-role-public   2       63484
  &LabelSelector{...}   default/fqdn                        1       39546
  &LabelSelector{...}   default/rule1                       1       4992
                                                                    39546
                                                                    63484
  &LabelSelector{...}   default/fqdn                        3       45238
  &LabelSelector{...}   default/fqdn                        1
  &LabelSelector{...}   default/from-world-to-role-public   1       9
                                                                    16777217
  &LabelSelector{...}   default/from-world-to-role-public   1       10
  &LabelSelector{...}   default/from-world-to-role-public   1       2
  api.github.com ...    default/fqdn                        1       16777217
  # cilium-dbg metrics list | grep selector
  cilium_fqdn_selectors                                     1.000000
  cilium_policy_selector_match_count_max   class="cluster"  0.000000
  cilium_policy_selector_match_count_max   class="fqdn"     1.000000
  cilium_policy_selector_match_count_max   class="other"    3.000000
  cilium_policy_selector_match_count_max   class="world"    2.000000

The cilium_fqdn_selectors metric counts the total number of fqdn
selectors altogether, but the cilium_policy_selector_match_count_max is
the maximum number of identities selected by any selector in the
specified class.

For now to avoid high cardinality, there are four categories:
- World: Equivalent to entity selector on world
- Cluster: Equivalent to entity selector on the cluster
- FQDN: Any selector that is based on an FQDNs statement
- Other: All other selectors.

"All" is notably omitted, because the Cilium dataplane already optimizes
such statements to minimize cardinality when encoding the identities into
policy maps.

Signed-off-by: Joe Stringer <joe@cilium.io>

christarazi reviewed (approved):
Nice!

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -136,6 +137,26 @@ func (sc *SelectorCache) GetModel() models.SelectorCache {
+func (sc *SelectorCache) Stats() selectorStats {
+	result := newSelectorStats()
+
+	sc.mutex.RLock()
+	defer sc.mutex.RUnlock()
```
I don't have a good sense for the impact of this. It's just the readlock, but how often does Prometheus call the `Collect()` function on a `Collector`? cc @cilium/metrics

chancez replied:
It occurs on each scrape, so it depends on your Prometheus scrape interval.

joestringer committed:
testutils: Consolidate DummySelectorCacheUser

No functional changes in this commit, just moving code.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy: Exclude subject selectors from metrics

Subject selectors don't have the same policymap implications as peer
selectors, so they are less interesting to track in selectorcache
cardinality metrics.

Signed-off-by: Joe Stringer <joe@cilium.io>

ianvernon reviewed (approved):
Nice work :)
